contrib/buildinfo/src/mill/contrib/buildinfo/BuildInfo.scala

```diff
@@ -1,58 +1,201 @@
+        buildInfoMembers(), isScala, buildInfoPackageName, buildInfoObjectName
+      )
+
+      val ext = if (isScala) "scala" else "java"
+
       os.write(
-        outputFile,
-        s"""|${buildInfoPackageName.map(packageName => s"package ${packageName}\n").getOrElse("")}
-            |object ${buildInfoObjectName} {
-            |$internalMembers
-            |
-            |  val toMap = Map[String, String](
-            |    $map)
-            |}""".stripMargin
+        T.dest / buildInfoPackageName.split('.') / s"${buildInfoObjectName}.$ext",
+        code,
+        createFolders = true
       )
-      (Seq(PathRef(outputFile)), PathRef(T.dest))
-    } else {
-      logger.debug("No build info member defined, skipping code generation")
-      (Seq.empty[PathRef], PathRef(T.dest))
+      Seq(PathRef(T.dest))
     }
   }
+}
 
-  override def generatedSources = T {
-    val (_, destPathRef) = generatedBuildInfo()
-    super.generatedSources() :+ destPathRef
+object BuildInfo{
+  def staticCompiledCodegen(buildInfoMembers: Map[String, String],
+                            isScala: Boolean,
+                            buildInfoPackageName: String,
+                            buildInfoObjectName: String): String = {
+    val bindingsCode = buildInfoMembers
+      .toSeq
+      .sortBy(_._1)
+      .map {
+        case (k, v) =>
+          if (isScala) s"""val $k = ${pprint.Util.literalize(v)}"""
+          else s"""public static java.lang.String $k = ${pprint.Util.literalize(v)};"""
+      }
+      .mkString("\n")
+
+
+    if (isScala) {
+      val mapEntries = buildInfoMembers
+        .map { case (name, _) => s""""$name" -> $name"""}
+        .mkString(",\n")
+
+      s"""
+         |package ${buildInfoPackageName}
+         |
+         |object $buildInfoObjectName {
+         |  $bindingsCode
+         |  val toMap = Map[String, String](
+         |    $mapEntries
+         |  )
+         |}
+      """.stripMargin.trim
+    } else {
+      val mapEntries = buildInfoMembers
+        .map { case (name, _) => s"""map.put("$name", $name);""" }
+        .mkString(",\n")
+
+      s"""
+         |package ${buildInfoPackageName};
+         |
+         |public class $buildInfoObjectName {
+         |  $bindingsCode
+         |
+         |  public static java.util.Map<String, String> toMap(){
+         |    Map<String, String> map = new HashMap<String, String>();
+         |    $mapEntries
+         |    return map;
+         |  }
+         |}
+      """.stripMargin.trim
+    }
   }
+  def codegen(buildInfoMembers: Map[String, String],
+              isScala: Boolean,
+              buildInfoPackageName: String,
+              buildInfoObjectName: String): String = {
+    val bindingsCode = buildInfoMembers
+      .toSeq
+      .sortBy(_._1)
+      .map {
+        case (k, v) =>
+          if (isScala) s"""val $k = this.readMillBuildInfo("$k")"""
+          else s"""public static java.lang.String $k = readMillBuildInfo("$k");"""
```

Comment:
This needs to be a method, not a field, otherwise the compiler is allowed to inline the value at the call side, which we need to avoid.

Reply:
> This needs to be a method, not a field, otherwise the compiler is allowed to inline the value at the call side, which we need to avoid.

I don't think that happens unless the RHS of the `final` field is a literal constant or (constant-folded). So in this case, a method call loading a JVM resource, it won't get inlined

Reply:
In the `buildInfoStaticCompiled = true` case, we need to make it a method, as the constants are candidates for inlining (https://docs.oracle.com/javase/specs/jls/se7/html/jls-15.html#jls-15.28), but that means we also need to make it a method in the other case, otherwise the generated code is not interchangeable.